### PR TITLE
Add tests for the supported query types.

### PR DESCRIPTION
**Pull request by jtibshirani**
The following query types should now be covered:
- `term`, `terms` and `terms_set`
- `prefix`
- `range`
- `match` and `multi_match`
- `query_string` and `simple_query_string`
- `exists`

I preferred unit tests as much as possible, but certain query types (like `match` and `query_string`) were more natural to check through integration tests.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java [hunk 1]
```diff
@@ -267,6 +267,21 @@ public Query existsQuery(QueryShardContext context) {
             return new PrefixQuery(term);
         }
 
+        @Override
+        public Query rangeQuery(Object lowerTerm,
+                                Object upperTerm,
+                                boolean includeLower,
+                                boolean includeUpper,
+                                QueryShardContext context) {
+            if (lowerTerm == null || upperTerm == null) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java]: Implemented naively, a range query with only one bound could incorrectly match values from other keys in the object. For example, a query with only a lower bound would become `(first\0value, null)`, and would match also values whose key is `second`.

We could fix this, but I wasn't sure the extra complexity was worth it. It's not clear to me how useful range queries will be right now, without support for numerics.

Comment (colings86) [reply server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java]: I feel like this limitation is ok for now. We can revisit this later if necessary

File: server/src/test/java/org/elasticsearch/index/mapper/KeyedJsonFieldTypeTests.java [hunk 2]
```diff
@@ -90,6 +120,26 @@ public void testFuzzyQuery() {
...
     }
 
+    public void testRangeQuery() {
+        KeyedJsonFieldType ft = createDefaultFieldType();
+        ft.setName("field");
+
+        TermRangeQuery expected = new TermRangeQuery("field",
+            new BytesRef("key\0lower"),
+            new BytesRef("key\0upper"), false, false);
+        assertEquals(expected, ft.rangeQuery("lower", "upper", false, false, null));
```
Comment (colings86) [inline server/src/test/java/org/elasticsearch/index/mapper/KeyedJsonFieldTypeTests.java]: should we also check that we do the right thing if the bounds are inclusive?

**Review by colings86 — COMMENTED** [review]
I left a comment but otherwise I think this is good. I'd like @romseygeek to take a look too though

File: server/src/test/java/org/elasticsearch/search/query/ExistsIT.java [hunk 3]
```diff
@@ -227,4 +228,41 @@ public void testFieldAliasWithNoDocValues() throws Exception {
...
+           .endObject()
+        .endObject();
+        assertAcked(prepareCreate("test").addMapping("type", mapping));
+
+        IndexRequestBuilder indexRequest = client().prepareIndex("test", "type", "1")
+           .setSource(XContentFactory.jsonBuilder()
+               .startObject()
+                   .startObject("headers")
+                       .field("content-type", "application/json")
+                   .endObject()
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/search/query/ExistsIT.java]: Not especially related to the current PR, but how is case normalization being handled?  The normal HTTP header here is `Content-type` with a capital C; is there a lowercase filter being applied anywhere, or is that being ignored for the moment?

**Review by romseygeek — APPROVED** [review]
LGTM.  I have one completely unrelated comment to do with case normalization, but feel free to answer that elsewhere...

File: server/src/test/java/org/elasticsearch/index/mapper/KeyedJsonFieldTypeTests.java (hunk 2 above)
Comment (jtibshirani) [reply server/src/test/java/org/elasticsearch/index/mapper/KeyedJsonFieldTypeTests.java]: 👍

**Pull request merged**